source/actors/player/PlayerHealth.hx

```diff
@@ -0,0 +1,141 @@
+        while (i >= 0) {
+            var heart:HeartContainer = healthGraphicGroup.members[i];
+            // test if damage exceeds current value of the heart
+            if (damage > heart.fragments) {
+                // damage exceeds value of heart, damage heart and keep going
+                damage -= heart.fragments;
+                heart.takeDamage(heart.fragments);
+                heart.updateGraphics();
+            } else {
+                // damage is less than value of heart, damage heart and break
+                heart.takeDamage(damage);
+                heart.updateGraphics();
+                break;
+            }
+            i--;
+        }
+
+        if (isDead()) {
+            trace("The player has died");
+        }
+    }
+
+    /**
+	 * Function to heal the player.
+     * @author Matt Lippelman
+     * @param amount an integer value in fragments (4 per heart) to heal the player
+     * @return void
+	 */
+    public function heal(amount:Int):Void {
+        for (val in 0...healthGraphicGroup.countLiving()) {
+            var heart:HeartContainer = healthGraphicGroup.members[val];
+            var missingFragments = HeartContainer.MAXFRAGAMOUNT - heart.fragments;
+            if (missingFragments == 0) {
+                continue;
+            } else if (amount > missingFragments) {
+                amount -= missingFragments;
+                heart.heal(missingFragments);
+                heart.updateGraphics();
+            } else {
+                heart.heal(amount);
+                heart.updateGraphics();
+                break;
+            }
+        }
+    }
+
+    /**
+	 * Function to increase maximum health by one full heart container.
+     * @author Matt Lippelman
+     * @return void
+	 */
+    public function increaseMaxHealthByOneHeart():Void {
+        increaseMaxHealth(4);
+    }
+
+    /**
+	 * Function to increase maximum health of the player.
+     * @author Matt Lippelman
+     * @param fragments an integer value in fragments (4 per heart) to be added to max health
+     * @return void
+	 */
+    public function increaseMaxHealth(fragments:Int):Void {
+        maxHealth += fragments;
+        healthGraphicGroup.add(new HeartContainer((healthGraphicGroup.countLiving()-1)*32, 0, 0));
+        fixGraphics(fragments);
+    }
+
+    /**
+	 * Helper function to fix graphics after increasing max health.
+     * @author Matt Lippelman
+     * @return void
+	 */
+    private function fixGraphics(fragments:Int):Void {
+        var health:Int = currentHealth() + fragments;
+        for (val in 0...healthGraphicGroup.countLiving()) {
+            var heart:HeartContainer = healthGraphicGroup.members[val];
+            if (health >= HeartContainer.MAXFRAGAMOUNT) {
+                health -= HeartContainer.MAXFRAGAMOUNT;
+                heart.setFragments(HeartContainer.MAXFRAGAMOUNT);
+            } else {
+                heart.setFragments(health);
+            }
+        }
+    }
```

Comment:
It makes sense to me that we might need to update graphics or fix things when we have a new heart container, but I think we could also leave the adjusting of health values and such to the Hero class rather than the display bar. Hero class could call `increaseMaxHealth` then call `heal` if they want to change the current health to increase in some way.